### PR TITLE
8332600: javac uses record components source position during compilation

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
javac uses the source position of record components to find and later probably remove a given record component during compilation. This could be necessary if annotation processors are present. This is done in part to provide better error messages for silly record definitions like:

record R(int i, float i) {} // two record components with the same name

but this is brittle and can backfire if the record is read from a class file as the source positions are not stored there. See [JDK-8332297](https://bugs.openjdk.org/browse/JDK-8332297) for some context

The idea of this fix is not to use the source position but only the name to find a given record component. This could imply that if the user defines an erroneous record like the one above, then the record could end up with less record components than expected but records like this wouldn't compile anyway

TIA

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8332600: javac uses record components source position during compilation`

### Issue
 * [JDK-8332600](https://bugs.openjdk.org/browse/JDK-8332600): javac uses record components source position during compilation (**Bug** - P4)

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20148/head:pull/20148` \
`$ git checkout pull/20148`

Update a local copy of the PR: \
`$ git checkout pull/20148` \
`$ git pull https://git.openjdk.org/jdk.git pull/20148/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20148`

View PR using the GUI difftool: \
`$ git pr show -t 20148`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20148.diff">https://git.openjdk.org/jdk/pull/20148.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20148#issuecomment-2223990801)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java [hunk 1]
```diff
@@ -1139,7 +1139,8 @@ public void visitMethodDef(JCMethodDecl tree) {
                         for (JCVariableDecl param: tree.params) {
                             boolean paramIsVarArgs = (param.sym.flags_field & VARARGS) != 0;
                             if (!types.isSameType(param.type, recordFieldTypes.head) ||
-                                    (recordComponents.head.isVarargs() != paramIsVarArgs)) {
+                                    (recordComponents.head != null &&
```
[the remark this answers is not in the document]
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java]: I have changed the code a bit, I think it make more sense now, thanks for the comments

**Review by lahodaj — APPROVED** [review]
Seems OK to me.

**Comment by vicente-romero-oracle** [comment]
thanks for the review

**Pull request closed without merging**